migrations/sql/schema/schema_1.0.0.py

```diff
@@ -0,0 +1,383 @@
+    ctx = Context()
+
+    instrument_name = ctx.get_mig_option("instrument")
+    repo = ctx.get_mig_option("butler-repo")
+    if not instrument_name or not repo:
+        raise ValueError(
+            "This migration script requires butler repository abd instrument name. "
+            "Please use `--options butler-repo=REPO --options instrument=NAME` command line options."
+        )
+
+    _check_insert_ids(ctx)
+    _check_flags()
+
+    cdVisitIds = _query_ccd_visit_ids(ctx)
+    _LOG.info("Found %s distinct ccdVisitIds", len(cdVisitIds))
+
+    # Find the (un)packer for this instrument.
+    packer = _make_packer(repo, instrument_name)
+
+    # Make and fill a table mapping ccdVisit to visit+detector
+    _make_map_table(ctx, cdVisitIds, packer)
+
+    # Add and fill visit + detector columns
+    _add_visit_detector_columns(ctx)
+
+    # Drop old column and update indices.
+    _drop_ccd_visit(ctx)
+
+    # Drop mapping table.
+    _LOG.info("Dropping %s table", ID_MAP_TABLE_NAME)
+    op.drop_table(ID_MAP_TABLE_NAME, schema=ctx.schema)
+
+    # Add all new flag columns to DiaSource and fill them.
+    _add_flag_columns(ctx)
+
+    # Drop old flags columns from all tables.
+    _drop_flag_columns(ctx)
+
+    # Update metadata version.
+    version = revision.split("_")[-1]
+    ctx.apdb_meta.update_tree_version("schema", version)
+
+
+def downgrade() -> None:
+    """Downgrade would be too complicated for this migration."""
+    raise NotImplementedError()
+
+
+def _check_insert_ids(ctx: Context) -> None:
+    """Check that we do not have any InsertId tables, those need to be
+    removed manually.
+    """
+    inspector = sqlalchemy.inspect(ctx.bind)
+    names = inspector.get_table_names(schema=ctx.schema)
+    _LOG.info("All table names: %s", names)
+    names = [name for name in names if name.endswith("InsertId")]
+    if names:
+        raise RuntimeError(
+            f"Schema contains InsertId tables, they have to be removed manually prior to migration: {names}"
+        )
+
+
+def _query_ccd_visit_ids(ctx: Context) -> set[int]:
+    """Make a list of all ccdVisitId values existing in the source tables."""
+    src = ctx.get_table("DiaSource")
+    fsrc = ctx.get_table("DiaForcedSource")
+
+    q1 = sqlalchemy.select(src.columns["ccdVisitId"]).distinct()
+    q2 = sqlalchemy.select(fsrc.columns["ccdVisitId"]).distinct()
+    result = ctx.bind.execute(q1.union(q2))
+    return set(result.scalars())
+
+
+def _make_packer(repo: str, instrument: str) -> Any:
+    """Find Butler dimensions packer."""
+    # This package does not depend on daf_butler, there is a chance that butler
+    # is not setup I do not want import of this module fail (alembic will be
+    # unhappy) so I delay import until this moment.
+    try:
+        from lsst.daf.butler import Butler
+        from lsst.obs.base import Instrument
+    except ImportError:
+        raise ImportError("Module lsst.obs.base cannot be imported, please setup obs_base.") from None
```

Comment:
Shouldn't this raise from, since you're hiding which of the imports failed (and it's corresponding error message)?

Reply:
The message says to setup obs_base, this should be sufficient as obs_base depends on daf_butler already. The order of imports makes it possible for daf_butler import to fail first, and I don't want to confuse people by saying `setup obs_base` when daf.butler import fails.